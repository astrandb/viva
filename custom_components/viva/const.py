"""Constants for the ViVa Weather integration."""

DOMAIN = "viva"
VERSION = "2025.8.0"
