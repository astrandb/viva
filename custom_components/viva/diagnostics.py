"""Diagnostics support for ViVa."""

from __future__ import annotations

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.core import HomeAssistant

from . import VivaConfigEntry

TO_REDACT: dict[str, Any] = {}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: VivaConfigEntry
) -> dict:
    """Return diagnostics for a config entry."""
    coordinator = config_entry.runtime_data.coordinator

    return {
        "config_entry_data": async_redact_data(dict(config_entry.data), TO_REDACT),
        "coordinator_data": coordinator.data,
    }
