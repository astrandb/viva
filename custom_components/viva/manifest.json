{
  "domain": "viva",
  "name": "ViVa Coastal Weather",
  "codeowners": [
    "@astrandb"
  ],
  "config_flow": true,
  "dependencies": [],
  "documentation": "https://github.com/astrandb/viva",
  "homekit": {},
  "integration_type": "service",
  "iot_class": "cloud_polling",
  "issue_tracker": "https://github.com/astrandb/viva/issues",
  "requirements": [],
  "ssdp": [],
  "version": "2025.8.0",
  "zeroconf": []
}
