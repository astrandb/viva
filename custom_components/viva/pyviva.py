"""pyviva - library for viva integration for Home Assistant."""

# Move to pypi.org when stable

from __future__ import annotations

import logging
from typing import Any, TypedDict

from aiohttp import ClientResponse, ClientResponseError, ClientSession

# import httpx

API_URL = "https://services.viva.sjofartsverket.se:8080/output/vivaoutputservice.svc/vivastation"

# Old web: http://vivadisplay.sjofartsverket.se/#/station/166

_LOGGER = logging.getLogger(__name__)


class StationModel(TypedDict):
    """Data model for Station."""

    ID: int
    Name: str
    Lat: float
    Lon: float


class Station:
    """Station class."""

    def __init__(self, raw_data: StationModel) -> None:
        """Init the class."""
        self.raw_data = raw_data

    @property
    def id(self) -> int:
        """Return id."""
        return self.raw_data["ID"]

    @property
    def name(self) -> str:
        """Return name."""
        return self.raw_data["Name"]

    @property
    def lat(self) -> float:
        """Return lattitude."""
        return self.raw_data["Lat"]

    @property
    def lon(self) -> float:
        """Return longitude."""
        return self.raw_data["Lon"]


class SampleModel(TypedDict):
    """Data model for sample observation."""

    Name: str
    Value: str
    Heading: int
    Unit: str
    Type: str
    Trend: str
    Msg: str
    Calm: int
    Updated: str
    StationID: int
    Quality: str
    WaterLevelReference: Any
    WaterLevelOffset: Any


class Observation:
    """Data model for observation."""

    def __init__(self, raw_data: SampleModel) -> None:
        """Init the class."""
        self.raw_data = raw_data

    @property
    def name(self) -> str:
        """Return name."""
        return self.raw_data["Name"]


class SingleStationObservation:
    """Class to get data from observation."""

    def __init__(self, raw_data: Observation) -> None:
        """Init the class."""
        self.raw_data = raw_data


class ViVaAPI:
    """Class to get data from ViVa API v1."""

    def __init__(self, websession: ClientSession) -> None:
        """Initialize."""
        self._websession = websession

    async def request(self, method, station="", **kwargs) -> ClientResponse:
        """Make a request."""
        headers = kwargs.get("headers")

        if headers is None:
            headers = {}
        else:
            headers = dict(headers)
            kwargs.pop("headers")

        res = await self._websession.request(
            method,
            f"{API_URL}/{station}",
            **kwargs,
            headers=headers,
        )
        res.raise_for_status()
        return res

    async def get_data(self):
        """Get data from api."""
        try:
            res = await self.request("GET")
            return await res.json()
        except ClientResponseError as exc:
            _LOGGER.error(
                "API get_data failed. Status: %s, - %s", exc.code, exc.message
            )

    async def get_all_stations(self) -> list[Station]:
        """Return all stations."""
        try:
            resp = await self.request("get", "")
            data = await resp.json()
            result = data["GetStationsResult"]["Stations"]
            return [Station(station_data) for station_data in result]
        except ClientResponseError as exc:
            _LOGGER.error(
                "API get_data failed. Status: %s, - %s", exc.code, exc.message
            )
        return []

    async def get_station(self, id_no: int) -> SingleStationObservation:
        """Return data from one station."""

        resp = await self.request("get", f"/{id_no}")
        data = await resp.json()
        result = data["GetSingleStationResult"]
        res = {}
        for sample in data["GetSingleStationResult"]["Samples"]:
            res[sample["Name"]] = sample
        result["Samples"] = res

        return result
