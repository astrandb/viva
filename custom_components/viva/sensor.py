"""Platform for sensor integration."""
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
import logging
from typing import Any

from homeassistant.components.sensor import (
    SensorDeviceClass,
    SensorEntity,
    SensorEntityDescription,
    SensorStateClass,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import UnitOfLength, UnitOfSpeed, UnitOfTemperature
from homeassistant.core import HomeAssistant
from homeassistant.helpers.device_registry import DeviceEntryType
from homeassistant.helpers.entity import DeviceInfo
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from . import get_coordinator
from .const import DOMAIN

SENSOR_TYPE_WIND = "wind"
SENSOR_TYPE_WIND_DIRECTION = "wind_direction"
SENSOR_TYPE_LEVEL = "level"
SENSOR_TYPE_WATER_TEMP = "watertemp"
SENSOR_TYPE_SIGHT = "sight"
SENSOR_TYPE_WAVE = "wave"
SENSOR_TYPE_WAVE_DIRECTION = "wave_direction"

_LOGGER = logging.getLogger(__name__)


@dataclass
class ViVaSensorDescription(SensorEntityDescription):
    """Class describing ViVa sensor entities."""

    type: str | None = None
    convert: Callable[[Any], Any] | None = None
    decimals: int = 1

WAVE_HEIGHT_SENSOR = ViVaSensorDescription(
    key="Våghöjd",
    type=SENSOR_TYPE_WAVE,
    device_class=SensorDeviceClass.DISTANCE,
    translation_key="wave_height",
    icon="mdi:waves",
    native_unit_of_measurement=UnitOfLength.METERS,
    state_class=SensorStateClass.MEASUREMENT,
)

WAVE_HEIGHT_DIRECTION_SENSOR = ViVaSensorDescription(
    key="Vågriktning",
    type=SENSOR_TYPE_WAVE_DIRECTION,
    icon="mdi:compass-outline",
    translation_key="wave_direction",
)

LEVEL_SENSOR = ViVaSensorDescription(
    key="Vattenstånd",
    type=SENSOR_TYPE_LEVEL,
    device_class=SensorDeviceClass.DISTANCE,
    translation_key="sealevel",
    icon="mdi:ferry",
    native_unit_of_measurement=UnitOfLength.CENTIMETERS,
    state_class=SensorStateClass.MEASUREMENT,
)

AVG_WIND_DIRECTION_SENSOR = ViVaSensorDescription(
    key="Vindriktning",
    type=SENSOR_TYPE_WIND_DIRECTION,
    icon="mdi:compass-outline",
    translation_key="wind_direction",
)

AVG_WIND_SENSOR = ViVaSensorDescription(
    key="Medelvind",
    type=SENSOR_TYPE_WIND,
    device_class=SensorDeviceClass.WIND_SPEED,
    translation_key="wind_strength",
    native_unit_of_measurement=UnitOfSpeed.METERS_PER_SECOND,
    suggested_unit_of_measurement=UnitOfSpeed.METERS_PER_SECOND,
    state_class=SensorStateClass.MEASUREMENT,
)

GUST_WIND_SENSOR = ViVaSensorDescription(
    key="Byvind",
    type=SENSOR_TYPE_WIND,
    device_class=SensorDeviceClass.WIND_SPEED,
    translation_key="gust_strength",
    native_unit_of_measurement=UnitOfSpeed.METERS_PER_SECOND,
    suggested_unit_of_measurement=UnitOfSpeed.METERS_PER_SECOND,
    state_class=SensorStateClass.MEASUREMENT,
)

TEMP_SENSOR = ViVaSensorDescription(
    key="Vattentemp",
    type=SENSOR_TYPE_WATER_TEMP,
    device_class=SensorDeviceClass.TEMPERATURE,
    translation_key="water_temperature",
    native_unit_of_measurement=UnitOfTemperature.CELSIUS,
    state_class=SensorStateClass.MEASUREMENT,
)

SIGHT_SENSOR = ViVaSensorDescription(
    key="Sikt",
    type=SENSOR_TYPE_SIGHT,
    device_class=SensorDeviceClass.DISTANCE,
    icon="mdi:binoculars",
    name="Sikt",
    native_unit_of_measurement="m",
)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the sensor platform."""
    coordinator = await get_coordinator(hass, config_entry)

    entities = []
    for obs in coordinator.data["Samples"]:
        obs2 = coordinator.data["Samples"][obs]
        if obs2["Type"] == SENSOR_TYPE_LEVEL:
            entities.append(ViVaSensor(coordinator, LEVEL_SENSOR, obs))
        elif obs2["Type"] == SENSOR_TYPE_WIND and obs2["Name"] == "Medelvind":
            entities.append(ViVaSensor(coordinator, AVG_WIND_SENSOR, obs))
            entities.append(ViVaSensor(coordinator, AVG_WIND_DIRECTION_SENSOR, obs))
        elif obs2["Type"] == SENSOR_TYPE_WIND and obs2["Name"] == "Byvind":
            entities.append(ViVaSensor(coordinator, GUST_WIND_SENSOR, obs))
        elif obs2["Type"] == SENSOR_TYPE_WATER_TEMP:
            entities.append(ViVaSensor(coordinator, TEMP_SENSOR, obs))
        elif obs2["Type"] == SENSOR_TYPE_SIGHT:
            entities.append(ViVaSensor(coordinator, SIGHT_SENSOR, obs))
        elif obs2["Type"] == SENSOR_TYPE_WAVE:
            entities.append(ViVaSensor(coordinator, WAVE_HEIGHT_SENSOR, obs))
            entities.append(ViVaSensor(coordinator, WAVE_HEIGHT_DIRECTION_SENSOR, obs))
        else:
            _LOGGER.warning(
                "Unsupported sensor type %s on station %s",
                obs2["Type"],
                obs2["StationID"],
            )
            continue
        _LOGGER.debug(
            "Setting up sensor %s on station %s", obs2["Type"], obs2["StationID"]
        )
    async_add_entities(entities)


class ViVaSensor(CoordinatorEntity, SensorEntity):
    """Representation of a Sensor."""

    entity_description: ViVaSensorDescription
    sensor_id: str = ""

    def __init__(self, coordinator, description: ViVaSensorDescription, sensor_id: str):
        """Initialize the sensor."""
        super().__init__(coordinator)
        self.entity_description = description
        self.sensor_id = sensor_id
        self._attr_has_entity_name = True
        self._attr_unique_id = (
            f"{self.coordinator.data['ID']}-{self.entity_description.key}"
        )
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, self.coordinator.data["ID"])},
            name=self.coordinator.data["Name"],
            manufacturer="Sjöfartsverket",
            model="ViVa",
            configuration_url="https://geokatalog.sjofartsverket.se/kartvisarefyren/",
            entry_type=DeviceEntryType.SERVICE,
        )

    @property
    def native_value(self) -> float | str | None:
        """Return the state of the sensor."""

        retstr = self.coordinator.data["Samples"][self.sensor_id].get("Value")

        if self.entity_description.type == SENSOR_TYPE_WAVE:
            retval = retstr.split()
            return retval[1]

        if self.entity_description.type == SENSOR_TYPE_WIND:
            retval = retstr.split()
            return retval[1]

        if self.entity_description.type == SENSOR_TYPE_WAVE_DIRECTION:
            retval = retstr.split()
            return retval[0]

        if self.entity_description.type == SENSOR_TYPE_WIND_DIRECTION:
            retval = retstr.split()
            return retval[0]

        if self.entity_description.type == SENSOR_TYPE_SIGHT:
            return retstr.replace(">", "")

        return retstr

    @property
    def available(self):
        """Return the availability of the entity."""

        return (
            self.coordinator.last_update_success
            and "Samples" in self.coordinator.data
            and self.sensor_id in self.coordinator.data["Samples"]
        )
