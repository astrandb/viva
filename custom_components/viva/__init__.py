"""The ViVa Weather integration."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from datetime import timedelta
import logging

from aiohttp import ClientResponseError

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import DOMAIN
from .pyviva import ViVaAPI

_LOGGER = logging.getLogger(__name__)

PLATFORMS = [Platform.SENSOR]

type VivaConfigEntry = ConfigEntry[VivaData]


@dataclass
class VivaData:
    """Runtime data for Viva integration."""

    api: ViVaAPI
    coordinator: DataUpdateCoordinator


async def async_setup_entry(hass: HomeAssistant, entry: VivaConfigEntry) -> bool:
    """Set up ViVa Weather from a config entry."""

    entry.runtime_data = VivaData(
        api=ViVaAPI(websession=async_get_clientsession(hass)), coordinator=None
    )

    coordinator = await get_coordinator(hass, entry)
    if not coordinator.last_update_success:
        await coordinator.async_config_entry_first_refresh()

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: VivaConfigEntry) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)


async def get_coordinator(
    hass: HomeAssistant,
    entry: VivaConfigEntry,
) -> DataUpdateCoordinator:
    """Get the data update coordinator."""
    if entry.runtime_data.coordinator:
        return entry.runtime_data.coordinator

    async def async_fetch():
        api = entry.runtime_data.api
        try:
            async with asyncio.timeout(10):
                return await api.get_station(entry.data["id"])
        except ClientResponseError as exc:
            _LOGGER.warning("API fetch failed. Status: %s, - %s", exc.code, exc.message)
            raise UpdateFailed(exc) from exc
        except TimeoutError as error:
            _LOGGER.warning("Timeout during coordinator fetch")
            raise UpdateFailed(error) from error

    entry.runtime_data.coordinator = DataUpdateCoordinator(
        hass,
        logging.getLogger(__name__),
        name=DOMAIN,
        update_method=async_fetch,
        update_interval=timedelta(minutes=5),
    )
    await entry.runtime_data.coordinator.async_refresh()
    return entry.runtime_data.coordinator
