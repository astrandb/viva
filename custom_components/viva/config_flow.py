"""Config flow for ViVa Weather integration."""

from __future__ import annotations

import logging
from typing import Any

import httpx
import voluptuous as vol

from homeassistant import config_entries
from homeassistant.core import HomeAssistant
from homeassistant.data_entry_flow import FlowResult
from homeassistant.exceptions import HomeAssistantError
from homeassistant.helpers.selector import (
    SelectOptionDict,
    SelectSelector,
    SelectSelectorConfig,
)

from .const import DOMAIN
from .pyviva import ViVaAPI

_LOGGER = logging.getLogger(__name__)

STEP_USER_DATA_SCHEMA = (
    vol.Schema(
        {
            vol.Required("id"): str,
        }
    ),
)


class ViVaHub:
    """Placeholder class to make tests pass.

    TODO Remove this placeholder class and replace with things from your PyPI package.
    """

    def __init__(self, host: str) -> None:
        """Initialize."""
        self.host = host

    async def authenticate(self, id_num: str) -> bool:
        """Test if we can authenticate with the host."""
        return True


async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
    """Validate the user input allows us to connect.

    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
    """

    hub = ViVaHub(data["id"])

    if not await hub.authenticate(data["id"]):
        raise InvalidAuth

    # If you cannot connect:
    # throw CannotConnect
    # If the authentication is wrong:
    # InvalidAuth

    # Return info that you want to store in the config entry.
    return {"title": "SjÃ¶fartsverket ViVa"}


class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
    """Handle a config flow for ViVa Weather."""

    VERSION = 1

    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> FlowResult:
        """Handle the initial step."""
        viva_api = ViVaAPI(websession=httpx.AsyncClient())
        station_list = [
            SelectOptionDict(value=str(st.id), label=st.name)
            for st in await viva_api.get_all_stations()
        ]

        if user_input is None:
            return self.async_show_form(
                step_id="user",
                data_schema=vol.Schema(
                    {
                        vol.Required("id"): SelectSelector(
                            SelectSelectorConfig(options=station_list)
                        ),
                    }
                ),
                errors={},
            )

        errors = {}
        await self.async_set_unique_id(user_input["id"])
        self._abort_if_unique_id_configured()

        try:
            await validate_input(self.hass, user_input)
            # info = await validate_input(self.hass, user_input)
        except CannotConnect:
            errors["base"] = "cannot_connect"
        except InvalidAuth:
            errors["base"] = "invalid_auth"
        except Exception:  # pylint: disable=broad-except
            _LOGGER.exception("Unexpected exception")
            errors["base"] = "unknown"
        else:
            title = next(
                (item for item in station_list if item["value"] == user_input["id"]),
                None,
            )["label"]
            return self.async_create_entry(title=title, data=user_input)

        return self.async_show_form(
            step_id="user",
            data_schema=vol.Schema(
                {
                    vol.Required("id"): SelectSelector(
                        SelectSelectorConfig(options=station_list)
                    ),
                }
            ),
            errors={},
        )


class CannotConnect(HomeAssistantError):
    """Error to indicate we cannot connect."""


class InvalidAuth(HomeAssistantError):
    """Error to indicate there is invalid auth."""
