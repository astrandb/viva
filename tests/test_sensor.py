"""Provide tests for viva sensors."""

from unittest.mock import patch

import pytest
from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
    snapshot_platform,
)
from syrupy import SnapshotAssertion

from custom_components.viva.const import DOMAIN
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_registry as er

from . import setup_integration
from .const import ENTRY_ID, MOCK_CONFIG


@pytest.mark.usefixtures("entity_registry_enabled_by_default")
async def test_sensor(
    hass: HomeAssistant,
    bypass_get_data,
    snapshot: SnapshotAssertion,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test sensor states."""

    mock_config_entry = MockConfigEntry(
        domain=DOMAIN, data=MOCK_CONFIG, entry_id=ENTRY_ID
    )

    with patch("custom_components.viva.PLATFORMS", [Platform.SENSOR]):
        await setup_integration(hass, mock_config_entry)

    await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)
