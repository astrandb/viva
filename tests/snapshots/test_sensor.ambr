# serializer version: 1
# name: test_sensor[sensor.forsmark_gust_strength-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
    }),
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.forsmark_gust_strength',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
      'sensor.private': dict({
        'suggested_unit_of_measurement': <UnitOfSpeed.METERS_PER_SECOND: 'm/s'>,
      }),
    }),
    'original_device_class': <SensorDeviceClass.WIND_SPEED: 'wind_speed'>,
    'original_icon': None,
    'original_name': 'Gust strength',
    'platform': 'viva',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'gust_strength',
    'unique_id': '12-Byvind',
    'unit_of_measurement': <UnitOfSpeed.METERS_PER_SECOND: 'm/s'>,
  })
# ---
# name: test_sensor[sensor.forsmark_gust_strength-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'wind_speed',
      'friendly_name': 'Forsmark Gust strength',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfSpeed.METERS_PER_SECOND: 'm/s'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.forsmark_gust_strength',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '4.0',
  })
# ---
# name: test_sensor[sensor.forsmark_sealevel-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
    }),
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.forsmark_sealevel',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <SensorDeviceClass.DISTANCE: 'distance'>,
    'original_icon': 'mdi:ferry',
    'original_name': 'Sealevel',
    'platform': 'viva',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'sealevel',
    'unique_id': '12-Vattenstånd',
    'unit_of_measurement': <UnitOfLength.CENTIMETERS: 'cm'>,
  })
# ---
# name: test_sensor[sensor.forsmark_sealevel-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'distance',
      'friendly_name': 'Forsmark Sealevel',
      'icon': 'mdi:ferry',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfLength.CENTIMETERS: 'cm'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.forsmark_sealevel',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '-33',
  })
# ---
# name: test_sensor[sensor.forsmark_water_temperature-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
    }),
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.forsmark_water_temperature',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <SensorDeviceClass.TEMPERATURE: 'temperature'>,
    'original_icon': None,
    'original_name': 'Water temperature',
    'platform': 'viva',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'water_temperature',
    'unique_id': '12-Vattentemp',
    'unit_of_measurement': <UnitOfTemperature.CELSIUS: '°C'>,
  })
# ---
# name: test_sensor[sensor.forsmark_water_temperature-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'temperature',
      'friendly_name': 'Forsmark Water temperature',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfTemperature.CELSIUS: '°C'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.forsmark_water_temperature',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '2.1',
  })
# ---
# name: test_sensor[sensor.forsmark_wind_bearing-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
    }),
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.forsmark_wind_bearing',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:compass-outline',
    'original_name': 'Wind bearing',
    'platform': 'viva',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'wind_bearing',
    'unique_id': '12-Vindriktning_int',
    'unit_of_measurement': '°',
  })
# ---
# name: test_sensor[sensor.forsmark_wind_bearing-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Forsmark Wind bearing',
      'icon': 'mdi:compass-outline',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': '°',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.forsmark_wind_bearing',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '101',
  })
# ---
# name: test_sensor[sensor.forsmark_wind_direction-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.forsmark_wind_direction',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:compass-outline',
    'original_name': 'Wind direction',
    'platform': 'viva',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'wind_direction',
    'unique_id': '12-Vindriktning',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensor[sensor.forsmark_wind_direction-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Forsmark Wind direction',
      'icon': 'mdi:compass-outline',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.forsmark_wind_direction',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'O',
  })
# ---
# name: test_sensor[sensor.forsmark_wind_strength-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
    }),
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.forsmark_wind_strength',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
      'sensor.private': dict({
        'suggested_unit_of_measurement': <UnitOfSpeed.METERS_PER_SECOND: 'm/s'>,
      }),
    }),
    'original_device_class': <SensorDeviceClass.WIND_SPEED: 'wind_speed'>,
    'original_icon': None,
    'original_name': 'Wind strength',
    'platform': 'viva',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'wind_strength',
    'unique_id': '12-Medelvind',
    'unit_of_measurement': <UnitOfSpeed.METERS_PER_SECOND: 'm/s'>,
  })
# ---
# name: test_sensor[sensor.forsmark_wind_strength-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'wind_speed',
      'friendly_name': 'Forsmark Wind strength',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfSpeed.METERS_PER_SECOND: 'm/s'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.forsmark_wind_strength',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '3.6',
  })
# ---
